### PR TITLE
added sending errors to Sentry

### PR DESCRIPTION
**Pull request by Danil42Russia**

### CODE REVIEW

File: src/main/kotlin/com/vk/admstorm/diagnostic/AdmStormLogger.kt [hunk 1]
```diff
@@ -31,5 +32,6 @@ class AdmStormLogger(delegateFactory: Factory, category: String) : DefaultLogger
 
     override fun error(message: String, t: Throwable?, vararg details: String?) {
         delegate.error(message, t, *details)
+        sentry.sendError(t)
```
Comment (i582) [inline src/main/kotlin/com/vk/admstorm/diagnostic/AdmStormLogger.kt:35]: Let's add this to `warn()` as well.

File: src/main/kotlin/com/vk/admstorm/diagnostic/SentryErrorReporter.kt [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+import com.intellij.util.Consumer
+import com.vk.admstorm.notifications.AdmNotification
+import com.vk.admstorm.services.SentryService
+import com.vk.admstorm.utils.MyUtils
+import io.sentry.protocol.SentryId
+import java.awt.Component
+
+class SentryErrorReporter : ErrorReportSubmitter() {
+    override fun getReportActionText(): String {
+        return "Report to Sentry"
```
Comment (i582) [inline src/main/kotlin/com/vk/admstorm/diagnostic/SentryErrorReporter.kt:20]: use `= "Report to Sentry"` syntax

File: src/main/kotlin/com/vk/admstorm/diagnostic/SentryErrorReporter.kt [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+    override fun submit(
+        events: Array<out IdeaLoggingEvent>,
+        additionalInfo: String?, parentComponent: Component,
+        consumer: Consumer<in SubmittedReportInfo>
+    ): Boolean {
+        val project = CommonDataKeys.PROJECT.getData(DataManager.getInstance().getDataContext(parentComponent))
+        if (project == null) {
+            consumer.consume(SubmittedReportInfo(SubmissionStatus.FAILED))
+            return false
```
Comment (i582) [inline src/main/kotlin/com/vk/admstorm/diagnostic/SentryErrorReporter.kt:31]: Maybe add here logging like `LOG.warn("cannot get project to submit exception to sentry"")`?

File: src/main/kotlin/com/vk/admstorm/diagnostic/SentryErrorReporter.kt [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+            }
+
+            val sentryId = sentry.sendError(event.data.throwable)
+            return if (sentryId != SentryId.EMPTY_ID) {
+                consumer.consume(SubmittedReportInfo(SubmissionStatus.NEW_ISSUE))
+                onSuccess(project, sentryId)
+                true
+            } else {
+                consumer.consume(SubmittedReportInfo(SubmissionStatus.FAILED))
+                false
```
Comment (i582) [inline src/main/kotlin/com/vk/admstorm/diagnostic/SentryErrorReporter.kt:47]: And here as well

File: src/main/kotlin/com/vk/admstorm/diagnostic/SentryErrorReporter.kt [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+            }
+        }
+
+        return true
+    }
+
+    private fun onSuccess(project: Project?, sentryId: SentryId) {
```
Comment (i582) [inline src/main/kotlin/com/vk/admstorm/diagnostic/SentryErrorReporter.kt:54]: project can be non-nullable

File: src/main/kotlin/com/vk/admstorm/git/GitUtils.kt [hunk 6]
```diff
@@ -562,6 +562,19 @@ object GitUtils {
         }
     }
 
+
+    private const val getUserNameCommand = "git -c core.quotepath=false config --local --null --get user.name"
```
Comment (i582) [inline src/main/kotlin/com/vk/admstorm/git/GitUtils.kt:566]: Please add here comment like the above other constants

File: src/main/kotlin/com/vk/admstorm/git/GitUtils.kt [hunk 7]
```diff
@@ -562,6 +562,19 @@ object GitUtils {
...
+    private const val getUserEmailCommand = "git -c core.quotepath=false config --local --null --get user.email"
+
+    data class User(val name: String?, val email: String?)
+
+    fun localUser(project: Project): User {
```
Comment (i582) [inline src/main/kotlin/com/vk/admstorm/git/GitUtils.kt:571]: And here as well

File: src/main/kotlin/com/vk/admstorm/git/GitUtils.kt [hunk 8]
```diff
@@ -562,6 +562,19 @@ object GitUtils {
         }
     }
 
+
+    private const val getUserNameCommand = "git -c core.quotepath=false config --local --null --get user.name"
+    private const val getUserEmailCommand = "git -c core.quotepath=false config --local --null --get user.email"
+
+    data class User(val name: String?, val email: String?)
```
Comment (i582) [inline src/main/kotlin/com/vk/admstorm/git/GitUtils.kt:569]: Are you sure these fields should be nullable?

File: src/main/kotlin/com/vk/admstorm/diagnostic/AdmStormLoggerFactory.kt [hunk 9]
```diff
@@ -1,11 +1,12 @@
 package com.vk.admstorm.diagnostic
 
 import com.intellij.openapi.diagnostic.Logger
+import com.vk.admstorm.services.SentryService
 
-class AdmStormLoggerFactory(private val factoryDelegate: Logger.Factory) : Logger.Factory {
+class AdmStormLoggerFactory(private val sentry: SentryService, private val factoryDelegate: Logger.Factory) : Logger.Factory {
     override fun getLoggerInstance(category: String): Logger {
         if (category.contains("com.vk.admstorm")) {
```
Comment (i582) [inline src/main/kotlin/com/vk/admstorm/diagnostic/AdmStormLoggerFactory.kt:8]: Maybe use  `SentryService.PLUGIN_ID` here?

File: src/main/kotlin/com/vk/admstorm/services/SentryService.kt [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+                event.contexts.setOperatingSystem(os)
+
+                val runtime = SentryRuntime().apply {
+                    name = application.versionName
+                    version = application.fullVersion
+                }
+                event.contexts.setRuntime(runtime)
+
+                event.user = User().apply {
+                    id = AdmStormSettingsState.getInstance().userNameForSentry
```
Comment (i582) [inline src/main/kotlin/com/vk/admstorm/services/SentryService.kt:53]: You need to add a `userNameForSentry` setting when connecting to server

**Review by i582 — CHANGES_REQUESTED** [review]

**Review by i582 — CHANGES_REQUESTED** [review]
Update CHANGELOG as well

File: src/main/kotlin/com/vk/admstorm/diagnostic/AdmStormLogger.kt (hunk 1 above)
Comment (Danil42Russia) [reply src/main/kotlin/com/vk/admstorm/diagnostic/AdmStormLogger.kt:36]: added

File: src/main/kotlin/com/vk/admstorm/diagnostic/SentryErrorReporter.kt (hunk 3 above)
Comment (Danil42Russia) [reply src/main/kotlin/com/vk/admstorm/diagnostic/SentryErrorReporter.kt:31]: We will use `LOG.info`

File: src/main/kotlin/com/vk/admstorm/diagnostic/SentryErrorReporter.kt (hunk 2 above)
Comment (Danil42Russia) [reply src/main/kotlin/com/vk/admstorm/diagnostic/SentryErrorReporter.kt]: fixed

File: src/main/kotlin/com/vk/admstorm/diagnostic/AdmStormLoggerFactory.kt (hunk 9 above)
Comment (Danil42Russia) [reply src/main/kotlin/com/vk/admstorm/diagnostic/AdmStormLoggerFactory.kt:8]: fixed

File: src/main/kotlin/com/vk/admstorm/git/GitUtils.kt (hunk 7 above)
Comment (Danil42Russia) [reply src/main/kotlin/com/vk/admstorm/git/GitUtils.kt:571]: fixed

File: src/main/kotlin/com/vk/admstorm/git/GitUtils.kt (hunk 6 above)
Comment (Danil42Russia) [reply src/main/kotlin/com/vk/admstorm/git/GitUtils.kt:566]: fixed

File: src/main/kotlin/com/vk/admstorm/services/SentryService.kt (hunk 10 above)
Comment (Danil42Russia) [reply src/main/kotlin/com/vk/admstorm/services/SentryService.kt:60]: added

**Review by i582 — APPROVED** [review]
LGTM

**Pull request merged**